src/proto/extension.rs

```diff
@@ -1,14 +1,81 @@
-use ssh_encoding::{Decode, Reader};
+//! SSH agent protocol extensions (messages & key constraints)
+
+use ssh_encoding::{CheckedSum, Decode, Encode, Error as EncodingError, Reader, Writer};
 use ssh_key::{public::KeyData, Signature};
 
-/// session-bind@openssh.com extension
+use super::ProtoError;
+
+// Reserved fields are marked with an empty string
+const RESERVED_FIELD: &str = "";
+
+/// SSH agent protocol *message* extension
+///
+/// Described in [draft-miller-ssh-agent-14 § 3.8](https://www.ietf.org/archive/id/draft-miller-ssh-agent-14.html#section-3.8)
+pub trait MessageExtension: 'static + Encode + Decode {
+    /// Extension name, indicating the type of the message (as a UTF-8 string).
+    ///
+    /// Extension names should be suffixed by the implementation domain
+    /// as per [RFC4251 § 4.2](https://www.rfc-editor.org/rfc/rfc4251.html#section-4.2),
+    fn extension_name() -> &'static str;
```

Comment:
```suggestion
    const NAME: &'static str;
```
Since we enforce the 'static lifetime, might as well make that a const.

(same below)